### PR TITLE
Exercise 2 awesome

### PR DESCRIPTION
**Pull request by menno4000**
trying to program in scala

### CODE REVIEW

File: src/main/scala/exercise2/Ior.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package exercise2
+
+sealed trait Ior[E, A] {
+  // todo: independent error type? => E_2?
```
Comment (KarakoA) [inline src/main/scala/exercise2/Ior.scala:4]: Yup, independent, yet related, error type here is best. :) One can do that with `Type Bounds`, we cover this in the upcoming lecture.

File: src/main/scala/exercise2/Ior.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  def flatMap[B](f: A => Ior[E, B]): Ior[E, B] = this match {
+    case Left(error) => Ior.left[E, B](error)
+    case Right(element) => f(element)
+    case Both(error, element) => f(element) match {
+      case l@Left(_) => l
```
Comment (KarakoA) [inline src/main/scala/exercise2/Ior.scala:9]: very nice!

File: src/main/scala/exercise2/SequentialCollections.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import scala.util.{Failure, Success, Try}
+
+trait StackLike[T] {
+  def top(): Option[T] = this match {
+    case StackElement(head, _) => Some(head)
+    case _ => None
+  }
+
+  def pop(): Try[StackLike[T]] = this match {
+    case StackElement(_, tail) => Success(tail)
```
Comment (KarakoA) [inline src/main/scala/exercise2/SequentialCollections.scala:13]: Neat!

File: src/main/scala/exercise2/SequentialCollections.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      case _ => this
+    }
+  }
+
+  def push(elem: T): StackLike[T] = StackElement[T](elem, this)
+}
+
+object StackLike {
+  def apply[T](elements: T*): StackLike[T] = {
+    val baseStack = StackEndElement[T]()
```
Comment (KarakoA) [inline src/main/scala/exercise2/SequentialCollections.scala:40]: can be done with `foldLeft` in one line: 
```scala
elements.foldLeft(StackEndElement[T]())((acc,c) =>acc.push(c))
```

File: src/main/scala/exercise2/SequentialCollections.scala [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  def enqueue(elem: T): QueueLike[T]
+
+  def dequeue(): Try[QueueLike[T]]
+
+  def front(): Option[T]
+
+  def isEmpty: Boolean
+}
+
+case class DoubleStackQueue[T]
```
Comment (KarakoA) [inline src/main/scala/exercise2/SequentialCollections.scala:68]: A trick one can use is to have the reversing of `in` in the constructor of the queue. This way it's guaranteed that the queue will have at least one element in out (if it has any elements at all) and will need less checks for that in `dequeue` and `front`.

File: src/main/scala/exercise2/KarplusStrongAlgorithm.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import scala.annotation.tailrec
+
+object KarplusStrongAlgorithm {
+
+  val EnergyDecayFactor = 0.996
+  val fundamentalFrequency = 440
+
+  def whiteNoise(frequency: Double = 440.0, volume: Double = 1.0): Queue[Double] = {
+    val realFrequency = (fundamentalFrequency * (fundamentalFrequency.toDouble/frequency))
```
Comment (KarakoA) [inline src/main/scala/exercise2/KarplusStrongAlgorithm.scala:15]: mm what's the idea behind fundamental frequency? `(1 to frequency )` should be alright

File: src/main/scala/exercise2/KarplusStrongAlgorithm.scala [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    val r = new Random()
+    (1 to realFrequency.toInt)
+      .map(_ => r.between(-0.5, 0.5) * volume)
+      .to(Queue)
```
Comment (KarakoA) [inline src/main/scala/exercise2/KarplusStrongAlgorithm.scala:19]: You are using scala's `immutable.Queue`, you should be using your own instead

**Review by KarakoA — APPROVED** [review]
Well done! `16/15` points!



- :star: For solving Task 1 A)
- :star: For solving Task 1 B)
- :star: For solving Task 1 C)
- :star: For solving Task 1 D)
- :star::star: For solving Task 2 A)
- :star::star: For solving Task 2 B)
- :star: For solving Task 3 A)
- :star: For solving Task 3 B)
- :star: For solving Task 3 C)
- :star: For solving Task 3 D)
- :star: For solving Task 3 E)
- :star: For not using `var` anywhere
- :star: For having tests
- :star2: for solving the optional task

File: src/main/scala/exercise2/KarplusStrongAlgorithm.scala (hunk 6 above)
Comment (menno4000) [reply src/main/scala/exercise2/KarplusStrongAlgorithm.scala:15]: We were confused by the definition behind 'frequency' in this context - usually, lower frequencies would result in 'lower' pitch for the produced sound, but the original method would produce a higher pitch for lower frequency input instead. We use a variable 'fundamental frequency' to invert the behaviour. We (probably wrongfully) borrowed the term from the [wikipedia article about the KS algorithm](https://en.wikipedia.org/wiki/Karplus%E2%80%93Strong_string_synthesis#Tuning_the_string), which incidentally also displays the issue of an inverse frequence-to-pitch relationship.

Comment (KarakoA) [reply src/main/scala/exercise2/KarplusStrongAlgorithm.scala:15]: Ahh, I see (and hear). So if one plays the "frequencies" `440` (A5) and `220` (A4), A5 should be higher in pitch, but it's actually lower with the default implementation.

**Pull request merged**